ranger/fs.go

```diff
@@ -0,0 +1,73 @@
+package ranger
+
+import (
+	"embed"
+	"errors"
+	"fmt"
+	"io/fs"
+	"sync"
+)
+
+// virtualFS implements fs.FS
+type virtualFS struct {
+	// A cache for minimizing ascertaining which directory holds the template.
+	cache map[string]func(string) (fs.File, error)
+
+	// The current working directory
+	osDir fs.FS
+
+	// Trails package-level directory
+	pkgDir fs.FS
+
+	sync.Mutex
+}
+
+// Open opens the file matching the name using the following strategy:
+// - check the cache
+// - check the OS filesystem
+// - check the package-level virtual filesystem
+//
+// Whenever a file is found and is not present in the cache, it is added.
+// Nothing removes references from the cache.
+//
+// The cache cannot become invalid at runtime since pkgDir is embedded.
+// If a file is removed from the OS during runtime,
+// then a reference to it from the cache returns the same error (fs.ErrNotExist)
+// as if the cache did not have that reference.
+func (vfs *virtualFS) Open(name string) (fs.File, error) {
+	// NOTE(dlk): while a concurrent routine could add a reference
+	// to the cache before this returns,
+	// let's err on the side of performance and not have this function
+	// blocking while waiting to read and only block when needing to write.
+	fn, ok := vfs.cache[name]
```

Comment:
For discussion: what about a cache only "filesystem" that uses similar logic and a slice of strings to load all templates into the cache at runtime?

Reply:
Did it! 🎉